Add channels index

KyleMaas opened the pull request:
Tried adding a channel index.  This actually seems to work pretty well.

arj03 commented:
Yes, nice. Need to test what this does for onboarding performance.

KyleMaas commented:
Without channel index:

```
full feeds: 513.962158203125 ms
partial feeds: 4963.927001953125 ms
```

With index:

```
full feeds: 548.82275390625 ms
partial feeds: 4259.909912109375 ms
```

KyleMaas commented:
Pulling channel list on Channels using Overall Popular (which queries all posts with a channel) vs. All Alphabetically (which uses the index):

```
channel list: 35.903076171875 ms
channel list: 0.638916015625 ms
```

This test was done with following both of your accounts so it would have more channel data to work with.

KyleMaas commented:
See https://github.com/arj03/ssb-browser-demo/pull/197 for uses

arj03 commented:
Yeah I think we can do this. We need a way to disable the index because not all browser apps want to use this index. I tested indexing and its 850 ms outside the browser so that seems ok. I have a few comments on the index that would be good to get in before we merge this.

arj03 commented on indexes/channels.js:
```diff
@@ -0,0 +1,109 @@
+const jsonCodec = require('flumecodec/json')
+const Plugin = require('ssb-db2/indexes/plugin')
+
+const isFeed = require('ssb-ref').isFeed
```
not needed

KyleMaas replied:
isFeed could do, but jsonCodec and Plugin are used further down.

arj03 replied:
Right, the comment was only for isFeed

arj03 commented on indexes/channels.js:
```diff
@@ -0,0 +1,109 @@
+
+const isFeed = require('ssb-ref').isFeed
+
+// 1 index:
```
just remove these 2 comment lines

arj03 commented on indexes/channels.js:
```diff
@@ -0,0 +1,109 @@
+  const { level, offset, stateLoaded, onData, writeBatch } = Plugin(
+    dir,
+    name,
+    3,
```
start with 1

arj03 commented on indexes/channels.js:
```diff
@@ -0,0 +1,109 @@
+  }
+  
+  function updateChannelData(channel) {
+    channels[channel] = { id: channel }
```
if you have a dict here, why not add a count? Might be useful to find most used channels?

KyleMaas commented:
Have not tested that version yet, but that should include all the changes.

KyleMaas commented:
Tried it, and it works pretty well.  Want to make one more change so that `##channel` is counted as just `channel`.  If we're going to keep counts, it makes a difference.

arj03 commented:
Yes, no rush with this. Lets get this index right before merging :)

arj03 commented on indexes/channels.js:
```diff
@@ -0,0 +1,113 @@
+  function updateChannelData(channel) {
+    if (!channels[channel])
+      channels[channel] = { id: channel, count: 0 }
+    if (!channels[channel].count) // For compatibility with existing index, if it exists.
```
I guess this is from your existing index? It's better if you just nuke it so we don't need that code here.

KyleMaas commented:
Fixed stray leading '#'s, removed compatibility code.  Checked it, and "patchbook" now shows a count of 3 where before you had to add "patchbook" and "#patchbook" together to get that.

KyleMaas commented:
I'm pretty happy with the way this works at this point.  Stripping out leading '#'s means I can remove some of the duplicate code from the channel-index branch of ssb-browser-demo.

arj03 commented:
Tested this. Did a full sync of my browser profile and the sync time did not noticeable change. Are the changes to ssb-browser-demo in the PR ready? It's no rush, just don't want to merge this before :)

KyleMaas commented:
Yes.  I left it in Draft status until this was ready, but it worked quite well when I last tested it.

arj03 commented:
Sweet. Will release this as 8.1.0 in a minute or two :)

KyleMaas commented:
Awesome.  You need me to update package.json and npm-shrinkwrap.json on the ssb-browser-demo branch to fix the merge conflicts, or do you want to do that?

arj03 commented:
Can you do that @KyleMaas ?
